Move IRSA (#2376)
giantswarm/docs: src/content/tutorials/access-management/iam-roles-for-service-accounts/index.md

---
linkTitle: IAM roles for service accounts
title: AWS IAM roles for service accounts (IRSA)
description: This article describes how to use a new feature that allows binding of specific AWS IAM roles to a service account of a pod.
weight: 60
menu:
  principal:
    parent: tutorials-access-management
    identifier: tutorials-access-management-irsa
user_questions:
 -  How can I use IAM roles for service accounts?
last_review_date: 2024-12-03
owner:
  - https://github.com/orgs/giantswarm/teams/team-phoenix
---

> *Basic knowledge about how service accounts work in Kubernetes is necessary for understanding what this page is about. If you could use a quick refresher, we strongly suggest taking a look at [this article](https://learnk8s.io/authentication-kubernetes).*
> 

Reviewer: ```suggestion
```

I don't think we should add so many details to a tutorial. The page already clearly explains that service accounts must be used.

You can associate an IAM role with a `Kubernetes` service account. This service account can then provide AWS permissions to the containers in any pod that uses that service account. With this feature, you no longer need to provide extended permissions to the Giant Swarm node IAM role so that pods on that node can call AWS APIs.

Applications must sign their AWS API requests with AWS credentials. This feature provides a strategy for managing credentials for your applications, similar to the way that Amazon EC2 instance profiles provide credentials to Amazon EC2 instances. Instead of creating and distributing your AWS credentials to the containers or using the Amazon EC2 instance’s role, you can associate an IAM role with a `Kubernetes` service account. The applications in the pod’s containers can then use an AWS SDK or the AWS CLI to make API requests to authorized AWS services.

> The official documentation from AWS: [IAM roles for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) `IRSA`.

Reviewer: ```suggestion
The official documentation from AWS: [IAM roles for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) (`IRSA`).
```

## IAM role permissions on `capa` controller

The Cluster APi for AWs controller comes with an attached policy name `giantswarm-<INSTALLATION>-irsa-controller-policy` that contains permissions like:

```json
  {
      "Effect": "Allow",
      "Action": [
        "iam:CreateOpenIDConnectProvider",
        ...
        "cloudfront:TagResource",
        ...
        "acm:RequestCertificate",
        ...
      ],
      "Resource": "*"
  },
  {
      "Effect": "Allow",
      "Action": [
        "s3:CreateBucket",
        ...
      ],
      "Resource": "arn:aws:s3:::*-g8s-*"
  }
```

This policy allows the controller to create a Cloudfront distribution, ACM certificate, S3 bucket and OpenID Connect provider to enable the usage of the IRSA feature.

> 

Reviewer: This shouldn't be here at all

## Using IAM roles for service accounts

### IAM role

> #### IAM Role Trusted Entity

Reviewer: Three levels of heading without clear guidance for the reader. Can we change this tutorial to the new expected form? That means: describe the goal in the beginning (use AWS SDK from an application pod), then one heading for each step (create IAM role, add service account annotation, show that it works).

Author: I am moving the content from vintage here, but I have not recreated the whole document. Let me see what I can do

The **trusted entity** of a IAM role needs to contain a condition like the `sub` service account name of a JSON Web token (JWT). You need to ensure one of those conditions match against the `JWT`, which is used by one of your applications.

> This is a example `JWT` from `external-dns` which is verified against the `external-dns` IAM role.

Reviewer: No need to go into technical details of JWTs at all. We need to give an example trust policy here that allows access by the Kubernetes service account.

Internal example:

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::242036376510:oidc-provider/irsa.golem.gaws.gigantic.io"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {
                    "irsa.golem.gaws.gigantic.io:sub": "system:serviceaccount:mimir:mimir-alertmanager"
                }
            }
        }
    ]
}
```

Also, we should state whether customers should rely on the OIDC provider domain that comes with the WC or create their own.

```json
{
  "aud": [
    "sts.amazonaws.com"
  ],
  "kubernetes.io": {
    "namespace": "kube-system",
    "pod": {
      "name": "external-dns-7dfdc97c55-v5spg",
    },
    "serviceaccount": {
      "name": "external-dns",
    }
  },
  "sub": "system:serviceaccount:kube-system:external-dns"
}
```

### Service account

The service account has to be annotated with a full `ARN` of the IAM role. You can get the `ARN` of the role in the AWS console, when you check role details.

The annotation key has to be `eks.amazonaws.com/role-arn`.

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    eks.amazonaws.com/role-arn: arn:aws:iam::$AWS_ACCOUNT:role/$IAM_ROLE_NAME
  name: s3-access
  namespace: default
```

## Cross Account Roles {#cross-account}

### IAM role cross account

In order to assume roles cross AWS accounts you will need to create a new AWS Identity Provider (OpenID Connect) in the AWS account where the IAM role is located.

![Creating AWS Identity Provider](identity-provider.png)

Login into the AWS on the account where the cluster is running:

> - Grab the address of the `Identity Provider` in your current cluster `IAM > Identity Providers`. It will look like `https://s3.REGION.amazonaws.com/AWS_ACCOUNT-g8s-CLUSTER_ID-oidc-pod-identity-v2` or `https://CLOUDFRONT_ID.cloudfront.net`.

Reviewer: ```suggestion
- Grab the address of the `Identity Provider` in your current cluster via [AWS Console > IAM > Identity Providers](https://us-east-1.console.aws.amazon.com/iam/home?#/identity_providers). It will look like `https://irsa.mycluster.<basedomain>`.
```

(for China, we can adapt this later)

> Login into the account where the IAM role is located and create an identity provider in `IAM > Identity Providers`:

Reviewer: ```suggestion
Log into the account where the IAM role is located and create an identity provider in `IAM > Identity Providers`:
```

- Set `Provider URL` to the previously gathered address and click the `Get thumbprint` to import the certificate.

> - Set the `audience` to `sts.amazonaws.com` OR `sts.amazonaws.com.cn` for China regions.

Reviewer: ```suggestion
- Set the `audience` to `sts.amazonaws.com` (for China regions: `sts.amazonaws.com.cn`).
```

## Verify your configuration is correct

Once your pod is running with the configured service account, you should see a file in the pod called `/var/run/secrets/eks.amazonaws.com/serviceaccount/token`, which contains a `JWT` token with details of the role.

> The pod should also have configured environment variables `AWS_WEB_IDENTITY_TOKEN_FILE` and `AWS_ROLE_ARN`.

Reviewer: ```suggestion
The pod should also automatically receive configured environment variables `AWS_WEB_IDENTITY_TOKEN_FILE` and `AWS_ROLE_ARN`.
```

Check the pod using command `kubectl -n NAMESPACE get pod POD_NAME -o yaml` and search for the environment variables or for the volume mounts.

> 

Reviewer: [optional] I think we should make this more actionable by deploying an example pod. Suggestion: aws-cli image with a shell in order to test some example IAM permission.

- Grab the address of the `Identity Provider` in your current cluster via [AWS Console > IAM > Identity Providers](https://us-east-1.console.aws.amazon.com/iam/home?#/identity_providers). It will look like `https://irsa.mycluster.<basedomain>`.

Log into the account where the IAM role is located and create an identity provider in `IAM > Identity Providers`:

- Set `Provider URL` to the previously gathered address and click the `Get thumbprint` to import the certificate.
- Set the `audience` to `sts.amazonaws.com` (for China regions: `sts.amazonaws.com.cn`).

## Verify your configuration is correct

Once your pod is running with the configured service account, you should see a file in the pod called `/var/run/secrets/eks.amazonaws.com/serviceaccount/token`, which contains a `JWT` token with details of the role.

The pod should also automatically receive configured environment variables `AWS_WEB_IDENTITY_TOKEN_FILE` and `AWS_ROLE_ARN`.

Check the pod using command `kubectl -n NAMESPACE get pod POD_NAME -o yaml` and search for the environment variables or for the volume mounts.

You can also use this example pod with `aws-cli` to verify the configuration:

> ```sh
> kubectl run aws-cli-test \
>   --image=amazon/aws-cli:2.13.17 \
>   --restart=Never \
>   --namespace=default \
>   --command -- sh -c "aws s3 ls" \
>   --overrides='{"spec": {"serviceAccountName": "default"}}'

Author: ```suggestion
```sh
kubectl run aws-cli-test \
  --image=amazon/aws-cli:2.13.17 \
  --restart=Never \
  --namespace=default \
  --command -- sh -c "aws s3 ls" \
  --overrides='{"spec": {"serviceAccountName": "default"}}'
```
```

The official documentation from AWS: [IAM roles for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) (`IRSA`).

## IAM role permissions on `capa` controller

The Cluster API for AWS controller comes with an attached policy name `giantswarm-<INSTALLATION>-irsa-controller-policy` that contains permissions like:

```json
  {
      "Effect": "Allow",
      "Action": [
        "iam:CreateOpenIDConnectProvider",
        "...",
        "cloudfront:TagResource",
        "...",
        "acm:RequestCertificate",
        "...",
      ],
      "Resource": "*"
  },
  {
      "Effect": "Allow",
      "Action": [
        "s3:CreateBucket",
        "...",
      ],
      "Resource": "arn:aws:s3:::*-g8s-*"
  }
```

Thanks to this policy, the controller can create a Cloudfront distribution, ACM certificate, S3 bucket and OpenID Connect provider to enable the usage of the IRSA feature.

## Using IAM roles for service accounts

To use IAM roles for service accounts, you need to create an IAM role, which permissions you want to assign to your application, and a `Kubernetes` service account.

### IAM role

> You have to make sure the **trusted entity** of the IAM role needs to contain a condition like the `sub` service account name of a JSON Web token (JWT). You need to ensure one of those conditions match against the `JWT`, which is used by one of your applications.

Reviewer 2: ```suggestion
You have to make sure the **trusted entity** of the IAM role contains a condition like the `sub` service account name of a JSON Web token (JWT). You need to ensure one of those conditions match against the `JWT`, which is used by one of your applications.
```

- Grab the address of the `Identity Provider` in your current cluster via [AWS Console > IAM > Identity Providers](https://us-east-1.console.aws.amazon.com/iam/home?#/identity_providers). It will look like `https://irsa.mycluster.<basedomain>`.

Log into the account where the IAM role is located and create an identity provider in `IAM > Identity Providers`:

- Set `Provider URL` to the previously gathered address and click the `Get thumbprint` to import the certificate.
- Set the `audience` to `sts.amazonaws.com` (for China regions: `sts.amazonaws.com.cn`).

## Verify your configuration is correct

Once your pod is running with the configured service account, you should see a file in the pod called `/var/run/secrets/eks.amazonaws.com/serviceaccount/token`, which contains a `JWT` token with details of the role.

The pod should also automatically receive configured environment variables `AWS_WEB_IDENTITY_TOKEN_FILE` and `AWS_ROLE_ARN`.

Check the pod using command `kubectl -n NAMESPACE get pod POD_NAME -o yaml` and search for the environment variables or for the volume mounts.

You can also use this example pod with `aws-cli` to verify the configuration:

> ```yaml
> apiVersion: v1
> kind: Pod
> metadata:
>   name: aws-cli-test
>   namespace: default # Change this to your desired namespace
> spec:
>   containers:
>   - name: aws-cli
>     image: amazon/aws-cli:2.13.17 # Latest version as of the last update
>     command: ["sh", "-c", "aws s3 ls"]
>   restartPolicy: Never
>   serviceAccountName: "" # Replace with the name of the service account you created
> ```

Author: ```suggestion
```sh
kubectl run aws-cli-test \
  --image=amazon/aws-cli:2.13.17 \
  --restart=Never \
  --namespace=default \
  --command -- sh -c "aws s3 ls" \
  --overrides='{"spec": {"serviceAccountName": "default"}}'
```

**Note**: You may want to set the correct `namespace` and `serviceAccountName`.
```

> ```

Reviewer 2: Maybe it's useful to aldo provide the kubectl one-liner version
```
kubectl run aws-cli-test \
  --image=amazon/aws-cli:2.13.17 \
  --restart=Never \
  --namespace=default \
  --command -- sh -c "aws s3 ls" \
  --overrides='{"spec": {"serviceAccountName": ""}}'
```
